stacks/resources/cognitoOidc/handlers/controller/linkedin.ts

```diff
@@ -0,0 +1,185 @@
+  const key = `${name.preferredLocale.language}_${name.preferredLocale.country}`;
+  return name.localized[key] || null;
+};
+const parseLocale = (name: any): string | null => {
+  if (!name?.preferredLocale?.country || !name?.preferredLocale?.language) return null;
+  return `${name.preferredLocale.language}-${name.preferredLocale.country}`;
+};
+
+const parseImageUrl = (imageElements: any[]): string | null => {
+  // find best image
+  // see https://docs.microsoft.com/en-us/linkedin/shared/references/v2/profile/profile-picture
+  if (!imageElements?.length) return null;
+  let bestRes = 0;
+  let best: any;
+  imageElements.forEach((ele) => {
+    const width = ele.data?.['com.linkedin.digitalmedia.mediaartifact.StillImage']?.storageSize?.width;
+    if (!width || width < bestRes) return;
+    bestRes = width;
+    best = ele;
+  });
+  if (!best) return null;
+  return best.identifiers?.[0]?.identifier || null;
+};
+
+interface LinkedInSecrets {
+  LINKEDIN_CLIENT_ID: string;
+  LINKEDIN_CLIENT_SECRET: string;
+}
+export const linkedin = (secrets: LinkedInSecrets) => {
+  const clientId = secrets.LINKEDIN_CLIENT_ID;
+  const redirUrl = process.env[ENV_COGNITO_REDIRECT_URI];
+  if (!redirUrl) throw new Error(`missing ${redirUrl} in env`);
+  return {
+    getAuthorizeUrl: (client_id: any, scope: any, state: any, response_type: any) => {
+      const newScope: string = filterOutScopesForLinkedin(scope);
+      const queryParameters = new URLSearchParams();
+      queryParameters.append('client_id', client_id);
+      queryParameters.append('scope', newScope);
+      queryParameters.append('state', state);
+      queryParameters.append('response_type', response_type);
+      queryParameters.append('redirect_uri', redirUrl);
+      return `${urls.oauthAuthorize}?${queryParameters.toString()}`;
+    },
+    getUserInfo: (accessToken: string) => {
+      console.debug('getUserInfo called');
+
+      return Promise.all([
+        getUserDetails<UserDetails>(accessToken).then((userDetails) => {
+          console.debug(`userDetails: ${JSON.stringify(userDetails, undefined, 4)}`);
+          const pictureElements = userDetails.profilePicture?.['displayImage~']?.elements || [];
+          return {
+            sub: userDetails.id,
+            name: `${userDetails.localizedFirstName} ${userDetails.localizedLastName}`,
+            vanityName: userDetails.vanityName || null,
+            picture: parseImageUrl(pictureElements),
+            locale: parseLocale(userDetails.firstName) || parseLocale(userDetails.lastName),
+            website: `https://www.linkedin.com/in/${userDetails.vanityName}`,
+
+            // custom attributes:
+            'custom:firstNameOriginal': parseOrigName(userDetails.firstName),
+            'custom:lastNameOriginal': parseOrigName(userDetails.lastName),
+            'custom:headline': userDetails.localizedHeadline,
+            'custom:vanityName': userDetails.vanityName,
```

Comment:
won't this force the use of `variable['custom:firstNameOriginal']` instead of being able to call it directly? or is it an OpenID thing? Might need a comment if so

Reply:
it's a mapping for cognito user attributes, used just for storing them in cognito